(HI-184) Inconsistent handling of fully-qualified facts when called from the command line [#puppethack]

rnelson0 opened the pull request:
Add a default_prof to the scope hash to create a matching no-colon-prepend key when a prepended key exists.

rnelson0 commented:
Thanks to @ripienaar for the assistance on this.

This change addresses HI-184. I found that rake -T did not show a spec target, though rake spec did work, so I am unsure what I would need to do to properly add a test to cover this. I'll look into it a bit and add to the PR if I find something. If there's a guide for this, or a clue why rake -T doesn't work, please let me know.

MikaelSmith commented:
Tried writing an acceptance test (https://github.com/puppetlabs/puppet/blob/master/docs/acceptance_tests.md). For some reason I can't reproduce the fixed behavior. The test is

```
test_name "Command-line lookup should resolve based on fact values from facter run"

agents.each do |agent|
  codedir = agent.puppet['codedir']
  hieradatadir = File.join(codedir, 'hieradata')

  teardown do
    apply_manifest_on agent, <<-PP
    file { '#{hieradatadir}':
      ensure  => directory,
      recurse => true,
      purge   => true,
      force   => true,
    }
    file { '#{codedir}/scope.yaml': ensure => absent }
    file { '#{codedir}/scope.json': ensure => absent }
    PP
  end

  osfamily = (on agent, facter('osfamily')).stdout.chomp

  apply_manifest_on agent, <<-PP
file { '#{codedir}/hiera.yaml':
  ensure  => present,
  content => '---
    :backends:
      - "yaml"
    :logger: "console"
    :hierarchy:
      - "osfamily/%{::osfamily}"
      - "%{environment}"
      - "global"

    :yaml:
      :datadir: "#{hieradatadir}"
  '
}

file { ['#{hieradatadir}', '#{hieradatadir}/osfamily']:
  ensure => directory,
}

file { '#{hieradatadir}/global.yaml':
  ensure => present,
  content => "---
    foo: zan
  "
}

file { '#{hieradatadir}/osfamily/#{osfamily}.yaml':
  ensure => present,
  content => "---
    foo: bar
  "
}
PP

  on agent, hiera('foo', "osfamily=#{osfamily}") do
    assert_match /bar/, result.stdout
  end

  on agent, hiera('foo', "::osfamily=#{osfamily}") do
    assert_match /bar/, result.stdout
  end

end
```

In my environment, I run with `be beaker --options-file merged_options.rb --hosts=config/nodes/debian-7-x86_64.yaml --preserve-hosts --tests=tests/test-184-command-line-facts.rb`.

rnelson0 commented:
Also ran the ruby suggested at https://gist.github.com/ripienaar/62b199d83e87fa0051d4 on my machine and in the beaker test nodes with no issues.

rnelson0 commented:
@MikaelSmith Here's a [gist](https://gist.github.com/rnelson0/a9de7d31c8983a8536cd) with the hiera.yaml, individual .yaml files, the tree layout (since gists won't let you path properly), and the output I receive. This is on CentOS release 6.4 (Final) using RVM with ruby 1.9.3p551 (2014-11-13 revision 48407) [x86_64-linux].

HAIL9000 commented:
Hey @rnelson0 have you gotten a chance to consult with @ripienaar about this acceptance test yet?

ripienaar commented:
I've unfortunately not had time to take a proper look

MikaelSmith commented:
I've setup a CentOS 6.6 docker machine with Ruby 1.9.3p551 using RVM, and this patch, and I still can't reproduce the fixed behavior @rnelson0 is seeing.

ripienaar commented:
OK, it's failing because https://github.com/puppetlabs/hiera/blob/master/lib/hiera/backend.rb#L311 checks if the particular key is ```include?```'d in the hash which our tweak wont address.

Why we could get it working here I dont have time to look at now but suffice to say we need to do some more fixing before this can be merged

whopper commented:
Hey @rnelson0! Just checking in to see if you'll have any time to continue on this.

rnelson0 commented:
@whopper I would love to but don't think I'll have time until after PuppetConf - or maybe at the contributor's summit there. Lots to do in ~3 weeks to prepare!

whopper commented:
@rnelson0 no worries! We'll probably see you at the contributor's summit too! :)

whopper commented:
Hey @rnelson0! Just checking in to see if you're still interested in working on this!

Hope you had a great Holiday!

rnelson0 commented:
Wow, time flies! I'm game for working on this if someone can help point me in the right direction for the beaker tests, which appears to highlight the issue. Is that something we can figure out on IRC?

whopper commented:
@rnelson0 actually we are having our PR triage tomorrow at 10:00AM PDX time, if you happen to be available to hang out and talk about the beaker issues for a little bit. If not, I'll put out an internal ping to see who'd have the most context to help you figure out what's going on, and when we'd have time to bring it in for work.

rnelson0 commented:
Unfortunately, tomorrow is no good for me. I'll be around either the 12th
or 19th if we can't find some other time to work it in before then.

whopper commented:
Mk - well I moved this PR to the 'discussion' section for tomorrow, so we'll talk about it and do proper triaging for where it'll land. Stay tuned!

MikaelSmith commented:
The acceptance test is just in the comment I put above. To run tests using Docker or Vagrant (using Docker with Beaker on Mac OS X is a pain)
- Change to `hiera/acceptance`
- Copy the test above into `./tests/test-184-command-line-facts.rb`
- Install gems: `bundle install --path=.bundle/gems`
- Create an centos7.yaml file containing:
```
# Docker
HOSTS:
  centos-7-64:
    platform: centos-7-x64
    image: centos:7
    hypervisor: docker
    roles:
    - agent
CONFIG:
  type: foss
```
or
```
# Vagrant
HOSTS:
  centos-7-64:
    platform: centos-7-x64
    box: puppetlabs/centos-7.2-64-nocm
    hypervisor: vagrant
    roles:
    - agent
CONFIG:
  type: foss
```
Then when neither of those work, start the Vagrant machine manually and use
```
HOSTS:
  centos-7-64:
    platform: centos-7-x64
    roles:
    - agent
CONFIG:
  type: foss
```
- Run tests with `bundle exec beaker --options config/git/options.rb --hosts centos7.yaml --tests tests/test-184-command-line-facts.rb --load-path lib --install "PUPPET/master,FACTER/2.x,https://github.com/rnelson0/hiera.git#normalize_cli_keys"`. You may need to specify `--keyfile ~/.ssh/a_key_file` as well, I'm not sure what the default behavior is.

You should see a failure along the lines of
```
Failed Tests Cases:
  Test Case tests/test-184-command-line-facts.rb reported: #<Minitest::Assertion: Expected /bar/ to match "zan\n".>
```

rnelson0 commented:
@MikaelSmith @ripienaar Very interesting. First, I added `tmp/setup` and `tmp/test` that can be used to set up some hiera files in `/var/hiera/` and then used for a test. You can check out that commit, I inserted it prior to the other changes, and then move forward commit by commit to see what happens.

On the PR triage today, we verified L311 looked like the culprit, so I inserted the `default_proc` at [L302](https://github.com/rnelson0/hiera/blob/normalize_cli_keys/lib/hiera/backend.rb#L302) and was not seeing any changes:
````
$ tmp/test

Basic test, no scope
expect Commonvalue: got Commonvalue

osfamily=Debian
expect Commonvalue: got Commonvalue

osfamily=RedHat
expect RedHatvalue: got Commonvalue

fqdn=other
expect Othervalue: got Commonvalue

fqdn=host.example.com
expect Commonvalue: got Commonvalue

::osfamily=Debian
expect Commonvalue: got Commonvalue

::osfamily=RedHat
expect RedHatvalue: got RedHatvalue

::fqdn=other
expect Othervalue: got Othervalue

::fqdn=host.example.com
expect Commonvalue: got Hostvalue
````
So I added some debugging around the original [L303](https://github.com/rnelson0/hiera/blob/normalize_cli_keys/lib/hiera/backend.rb#L303), resulting in:
````
      def qualified_lookup(segments, hash)
        value = hash
        value.default_proc = ->(hash, key) {        #L303
          if key =~ /^::(.+)/
            otherkey =$1
          else
            otherkey = "::#{key}"
          end

          if hash.include?(otherkey)
            hash[key] = hash[otherkey]
          else
            hash[otherkey] = hash[key]        #L313
          end
        }

        segments.each do |segment|
          if segment =~ /osfam/
            puts ""
            puts "value of segment: #{segment}"
            puts "value of osfamily: #{value['osfamily']}"
            puts "value of ::osfamily: #{value['::osfamily']}"
          end
          throw :no_such_key if value.nil?
          ...
````
Here's where it gets interesting:
````
$ tmp/test

Basic test, no scope
expect Commonvalue: got
value of segment: ::osfamily
/home/rnelson0/git/hiera/lib/hiera/backend.rb:303: stack level too deep (SystemStackError)

osfamily=Debian
expect Commonvalue: got
value of segment: ::osfamily
value of osfamily: Debian
value of ::osfamily: Debian
Commonvalue

osfamily=RedHat
expect RedHatvalue: got
value of segment: ::osfamily
value of osfamily: RedHat
value of ::osfamily: RedHat
RedHatvalue

fqdn=other
expect Othervalue: got
value of segment: ::osfamily
/home/rnelson0/git/hiera/lib/hiera/backend.rb:313: stack level too deep (SystemStackError)

fqdn=host.example.com
expect Commonvalue: got
value of segment: ::osfamily
/home/rnelson0/git/hiera/lib/hiera/backend.rb:304: stack level too deep (SystemStackError)

::osfamily=Debian
expect Commonvalue: got
value of segment: ::osfamily
value of osfamily: Debian
value of ::osfamily: Debian
Commonvalue

::osfamily=RedHat
expect RedHatvalue: got
value of segment: ::osfamily
value of osfamily: RedHat
value of ::osfamily: RedHat
RedHatvalue

::fqdn=other
expect Othervalue: got
value of segment: ::osfamily
/home/rnelson0/git/hiera/lib/hiera/backend.rb:303: stack level too deep (SystemStackError)

::fqdn=host.example.com
expect Commonvalue: got
value of segment: ::osfamily
/home/rnelson0/git/hiera/lib/hiera/backend.rb:313: stack level too deep (SystemStackError)
````
Lines 303/313 are marked above. I'm not sure why, but adding debugging appears to introduce a heisenbug. If you run this a few times, you'll see that the errors are random. Eventually I got `tmp/test: line 42:  2984 Segmentation fault      $HIERA key ::fqdn=other` on some lookups. Sometimes it totally goes out to lunch and I have to CTRL-Z; kill %1 to end it.

rnelson0 commented:
OK, some more poking at it makes me think I got it. Rather than changing `default_proc` I simply made sure the "other" key is available. The docker test passes now. I'm not sure the original fix is required, then.

MikaelSmith commented:
Probably don't want to add all the files in `tmp/`.

MikaelSmith commented on lib/hiera/backend.rb:
```diff
@@ -300,8 +300,19 @@ def clear!
+	    value[segment] = value[othersegment]
+	  end
+
+	  throw :no_such_key if value.nil?
```
You want this before value.include?, or you'll cause an error to be thrown about value being nil.

rnelson0 commented:
@MikaelSmith I'll remove that when I rebase it.

rnelson0 commented:
@MikaelSmith I think commit 52d63cc is probably not needed, and in fact doesn't seem to affect anything. Of course the two temporary commits can go and then the remaining three can be squashed down to one. Before I whack that commit, do you see any reason to keep it around?

MikaelSmith commented on bin/hiera:
```diff
@@ -243,6 +243,19 @@ unless options[:verbose]
   Hiera.logger = "noop"
 end
 
+# Allow lookup of keys with or without prepended double colons
```
This doesn't seem necessary to fix the issue we've seen. I don't understand how the scope is used sufficiently to know whether there's a case we haven't thought of, but I lean towards doing the minimal fix until we come up with a case that needs this.

rnelson0 commented:
@hlindberg We discussed this on the Core Triage today and this is ready for your review now! Let me know if you have questions. Thanks!

hlindberg commented:
I have problems with the change. It is the command line tool that seems to be at fault and not the hiera backend which I think should remain unchanged. I based this on the problem description in the ticket.

The tests are also failing.

I am also wondering if any of this is worth the time/trouble in the first place. (I.e. making changes to the hiera command line tool) since the new `puppet lookup` command works, is what will be used in the future, and has more options than the old hiera commad line tool (which is broken in several other ways as well).

MikaelSmith commented on spec/unit/backend_spec.rb:
```diff
@@ -212,12 +212,12 @@ def lookup(key, scope, order_override, resolution_type)
       @leading_double_colon_tests = {
-        "test_%{::rspec::data}_test" => "test__test",
-        "test_%{scope('::rspec::data')}_test" => "test__test"
+        "test_%{::rspec::data}_test" => "test_value_test",
```
I agree with Henrik, changing this test looks strange.

MikaelSmith commented on lib/hiera/backend.rb:
```diff
@@ -307,8 +307,19 @@ def clear!
+
         segments.each do |segment|
           throw :no_such_key if value.nil?
+          if segment =~ /^::(.+)/
```
Now that I think about it, we should be able to fix this from `bin/hiera`. Going to try just adding entries to options[:scope] rather than overriding default_proc.

MikaelSmith replied:
Putting
```
options[:scope].keys.each do |key|
  if key =~ /^::(.+)/
    otherkey = $1
  else
    otherkey = "::%s" % key
  end
  options[:scope][otherkey] = options[:scope][key]
end
```
in bin/hiera works. `default_proc` doesn't appear to be called on `include?`.

MikaelSmith commented on acceptance/ubuntu.yaml:
```diff
@@ -0,0 +1,10 @@
+# Docker
```
Remove this from the final commit.

MikaelSmith commented on acceptance/tests/test-184-command-line-facts.rb:
```diff
@@ -0,0 +1,66 @@
+}
+PP
+
+  on agent, hiera('foo', "osfamily=#{osfamily}") do
```
It would be nice to extend the acceptance test with all the test cases from the `tmp` dir. Then remove the `tmp` dir.

hlindberg commented:
As I said earlier and on the ticket - is it really worth spending time on this? A usability feature on a semi broken command? Remember that It will be bundled with puppet versions where the lookup command exists, and it works with hiera data, environment data and data in modules, the hiera command line tool only with global hiera data. I rather fix issues in lookup if there is something lookup cannot do. We have invested a lot of development time and QA time to ensure it works well - the hiera command line tool is in stark contrast; there is always something wrong with it.

rnelson0 commented:
If this fix can only be ported to hiera 3, then it appears to be a moot point. If it can be ported to hiera 1.x, then admins still using puppet 3 may rejoice.

hlindberg commented:
Even if i is possible to backport a fix, we are not planning to release any new versions except if there are security issues. We will do a 3.0.7 shortly. Release of 3.1.0 is not yet determined, and this PR currently targets master which is for  3.1.0.

rnelson0 commented:
@hlindberg With that in mind, if you think `puppet lookup` is the best place to focus this, it sounds like the PR should be closed and that you're inclined to close the HI-184 ticket, too. I am okay with that, I learned a lot anyway. I'll have to spend some time with `puppet lookup` soon and make sure things work right there!

hlindberg commented:
@rnelson0 yes, I think the energy is best spent on the new lookup implementation (command line tool, and function) for which we have a substantial testing suite, proper documentation and APIs. It may not be perfect yet as it is not tested i battle, and it may not support all the wanted use cases. But we want to hear about those and add them.

joshcooper commented:
@rnelson0 Thanks for working on this PR. It sounds like you and @hlindberg have a path forward using `puppet lookup` so I'm going to close this PR and ticket. It would be great to get your feedback with `lookup` and make sure it satisfies the use cases you care about!

rnelson0 commented:
@joshcooper will do!
